inifile.cc

```diff
@@ -44,98 +46,55 @@ typedef struct inifile_section_s {
 #define DELTA_BUFSIZE 128
 
 #define GPSBABEL_INIFILE "gpsbabel.ini"
+#define GPSBABEL_SUBDIR ".gpsbabel"
 
 /* Remember the filename we used so we can include it in errors. */
-char* gbinipathname;
+QString gbinipathname;
 
-static char*
-find_gpsbabel_inifile(const char* path)		/* can be empty or NULL */
+static QString
+find_gpsbabel_inifile(const QString& path)  /* can be empty or NULL */
 {
-  FILE* test;
-  char* buff;
-  int len;
-
-  if (path == NULL) {
-    return NULL;
-  }
-
-  len = strlen(path);
-  buff = (char*) xmalloc(len + 1 + strlen(GPSBABEL_INIFILE) + 1);
-  strcpy(buff, path);
-  if (len > 0) {
-    char test = buff[len - 1];
-#ifdef __WIN32__
-    if ((test != '\\') && (test != ':')) {
-      strcat(buff, "\\");
-    }
-#else
-    if (test != '/') {
-      strcat(buff, "/");
-    }
-#endif
-  }
-  strcat(buff, GPSBABEL_INIFILE);
-  test = fopen(buff, "rb");
-  if (test) {
-    fclose(test);
-    return buff;
+  if (path.isNull()) {
+    return QString();
   }
-  xfree(buff);
-  return NULL;
+  QString inipath(QDir(path).filePath(GPSBABEL_INIFILE));
+  return QFile(inipath).open(QIODevice::ReadOnly) ? inipath : QString();
 }
 
 static gbfile*
 open_gpsbabel_inifile()
 {
-  char* name;
-  char* envstr;
+  QString name;
+  QString envstr;
   gbfile* res = NULL;
 
-  envstr = getenv("GPSBABELINI");
-  if (envstr != NULL) {
-    FILE* test;
-
-    test = fopen(envstr, "r");
-    if (test != NULL) {
-      fclose(test);
+  envstr = ugetenv("GPSBABELINI");
```

Comment:
Sinze the FREEZE_TIME stuff is only used by our test framework to set "current time" to 1-1-1970, and not really an externally documented/useful behaviour, this is really the only place where getenv returning funky characters is "real", right?

Reply:
The most plausible example of how you can get environment variables with weird characters is if the Windows username is non-ASCII, which seems a common enough scenario in non-English locales. Presumably the Windows user directory is then "C:\Users\%USERNAME%" (unless Windows instituted some workaround that hashes unicode usernames into ASCII, which wouldn't surprise me at all, 8.3-hashed filenames anybody?), and any "INI file path" based off of that, would also have non-ASCII characters.